[Data] [Docs] Consolidate shuffling-related information into `Shuffling Data` page (#44098)
ray-project/ray: doc/source/data/shuffling-data.rst

.. _shuffling_data:

==============
Shuffling Data
==============

When consuming or iterating over Ray :class:`Datasets <ray.data.dataset.Dataset>`, it can be useful to
shuffle or randomize the order of data (for example, randomizing data ingest order during ML training). 
This guide shows several different methods of shuffling data with Ray Data and their respective trade-offs.

Types of shuffling
==================

Ray Data provides several different options for shuffling data, trading off the granularity of shuffle
control with memory consumption and runtime. The options below are listed in increasing order of 
resource consumption and runtime; choose the most appropriate method for your use case.

.. _shuffling_file_order:

Shuffle the ordering of files
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To randomly shuffle the ordering of input files before reading, call a function like
:func:`~ray.data.read_images` and specify ``shuffle="files"``. This randomly assigns
input files to workers for reading.

.. testcode::

    import ray

    ds = ray.data.read_images(
        "s3://anonymous@ray-example-data/image-datasets/simple",
        shuffle="files",
    )

> .. tip::

Reviewer: This didn't really feel like a tip to me. IMO, it might be better to make this regular text here and in the other sections. In general, I think we should try to use admonitions sparingly.

    This is the fastest option for shuffle, and is a purely metadata operation. This
    option doesn't shuffle the actual rows inside files, so the randomness might be
    poor if each file has many rows.

.. _local_shuffle_buffer:

Local shuffle when iterating over batches
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To locally shuffle a subset of rows, call a function like :meth:`~ray.data.Dataset.iter_batches`
and specify `local_shuffle_buffer_size`. This shuffles the rows up to a provided buffer
size during iteration. See more details in

> :ref:`Iterating over batches with shuffling <iterating-over-batches-with-shuffling>`.

Reviewer 2: Should we move that information to this page and then have a small reference on that page to this broader shuffle page?

.. testcode::

    import ray

    ds = ray.data.read_images("s3://anonymous@ray-example-data/image-datasets/simple")

    for batch in ds.iter_batches(
        batch_size=2,
        batch_format="numpy",
        local_shuffle_buffer_size=250,
    ):
        print(batch)

.. tip::

    This is slower than shuffling ordering of files, and shuffles rows locally without
    network transfer. This local shuffle buffer can be used together with shuffling
    ordering of files; see :ref:`Shuffle the ordering of files <shuffling_file_order>`.

>     If you observe reduced throughput when using ``local_shuffle_buffer_size``;
>     one way to diagnose this is to check the total time spent in batch creation by
>     examining the ``ds.stats()`` output (``In batch formatting``, under
>     ``Batch iteration time breakdown``).

Reviewer: ```suggestion
    If you observe reduced throughput when using ``local_shuffle_buffer_size``,
    check the total time spent in batch creation by
    examining the ``ds.stats()`` output (``In batch formatting``, under
    ``Batch iteration time breakdown``).
```

>     ``Batch iteration time breakdown``).
>     
>     If this time is significantly larger than the

Reviewer: Seems like these sentences should be part of the same paragraph?
```suggestion
    ``Batch iteration time breakdown``). If this time is significantly larger than the
```

>     time spent in other steps, one way to improve performance is to decrease
>     ``local_shuffle_buffer_size`` or turn off the local shuffle buffer altogether and only :ref:`shuffle the ordering of files <shuffling_file_order>`.

Reviewer: ```suggestion
    time spent in other steps, decrease
    ``local_shuffle_buffer_size`` or turn off the local shuffle buffer altogether and only :ref:`shuffle the ordering of files <shuffling_file_order>`.
```

Shuffle all rows
~~~~~~~~~~~~~~~~

To randomly shuffle all rows globally, call :meth:`~ray.data.Dataset.random_shuffle`.

.. testcode::

    import ray

    ds = (
        ray.data.read_images("s3://anonymous@ray-example-data/image-datasets/simple")
        .random_shuffle()
    )

.. tip::

    This is the slowest option for shuffle, and requires transferring data across
    network between workers. This option achieves the best randomness among all options.

.. _optimizing_shuffles:

> Advanced: Optimizing shuffles

Reviewer 2: Might just be a formatting thing, but should this be a subheading or a top level heading? Actually seems like all of the subsections are subsections of "Types of shuffling", is that intentional?

-----------------------------

> Shuffle operations are *all-to-all* operations where the entire Dataset must be materialized in memory before execution can proceed.

Reviewer 2: This section seems to not be super relevant to shuffling? Is the idea that these optimization might also apply to other all-to-all operations? The "these" in the below line is also unclear. I would have thought it was talking about shuffle operations but think it is talking about all to all operations?

Reviewer 2: Maybe move some of this to the "Enabling push-based shuffle" below which seems related?

Reviewer: +1 seemed slightly out of place to me. Wonder if we should just remove this section?

Currently, these are:

* :meth:`Dataset.groupby <ray.data.Dataset.groupby>`
* :meth:`Dataset.random_shuffle <ray.data.Dataset.random_shuffle>`
* :meth:`Dataset.sort <ray.data.Dataset.sort>`

.. note:: This is an active area of development. If your Dataset uses a shuffle operation and you are having trouble configuring shuffle, 
    `file a Ray Data issue on GitHub <https://github.com/ray-project/ray/issues/new?assignees=&labels=bug%2Ctriage%2Cdata&projects=&template=bug-report.yml&title=[data]+>`_. 

When should you use global per-epoch shuffling?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Use global per-epoch shuffling only if your model is sensitive to the

> randomness of the training data. Based on a
> `theoretical foundation <https://arxiv.org/abs/1709.10432>`__ all

Reviewer 2: ```suggestion
randomness of the training data. Based on a
`theoretical foundation <https://arxiv.org/abs/1709.10432>`__ , all
```

The increase compounds with distributed data-parallel training on a multi-node cluster due
to data transfer costs. This cost can be prohibitive when using very large datasets.

The best route for determining the best tradeoff between preprocessing time and cost and
per-epoch shuffle quality is to measure the precision gain per training step for your
particular model under different shuffling policies:

* no shuffling,
* local (per-shard) limited-memory shuffle buffer,
* local (per-shard) shuffling,
* windowed (pseudo-global) shuffling, and
* fully global shuffling.

As long as your data loading and shuffling throughput is higher than your training throughput, your GPU should
be saturated. If you have shuffle-sensitive models, push the
shuffle quality higher until this threshold is hit.

.. _shuffle_performance_tips:

Enabling push-based shuffle
~~~~~~~~~~~~~~~~~~~~~~~~~~~

Some Dataset operations require a *shuffle* operation, meaning that data is shuffled from all of the input partitions to all of the output partitions.
These operations include :meth:`Dataset.random_shuffle <ray.data.Dataset.random_shuffle>`,

> :meth:`Dataset.sort <ray.data.Dataset.sort>` and :meth:`Dataset.groupby <ray.data.Dataset.groupby>`.

Reviewer 2: It's not super intuitive why these operations require a shuffle, if possible maybe add a quick sentence explaining?

> Shuffle can be challenging to scale to large data sizes and clusters, especially when the total dataset size can't fit into memory.

Reviewer 2: ```suggestion
Shuffling can be challenging to scale to large data sizes and clusters, especially when the total dataset size can't fit into memory.
```

> Datasets provides an alternative shuffle implementation known as push-based shuffle for improving large-scale performance.

Reviewer 2: Think this is outdated verbage?
```suggestion
Ray Data provides an alternative shuffle implementation known as push-based shuffle for improving large-scale performance.
```
